File: tests/php/golden/SerializationTest.php
<?php

declare(strict_types=1);

namespace MoySklad\Tests\Golden;

use MoySklad\Tests\TestCase;

/**
 * Golden тесты для проверки сериализации/десериализации моделей SDK.
 * 
 * Тест выполняет roundtrip проверку:
 * 1. Загружает эталонный JSON из fixtures
 * 2. Десериализует JSON в объект модели SDK
 * 3. Сериализует объект обратно в JSON
 * 4. Сравнивает результат с эталоном
 * 
 * Это гарантирует, что SDK корректно обрабатывает все поля сущностей
 * без потери или искажения данных.
 * 
 * @see https://dev.moysklad.ru/doc/api/remap/1.2/
 */
class SerializationTest extends TestCase
{
    /**
     * Маппинг fixture файлов на классы моделей SDK.
     * Ключ - имя файла без расширения, значение - короткое имя класса модели.
     */
    private const FIXTURE_MODEL_MAP = [
        'product' => 'Product',
        'product_full' => 'Product',
        'product_image' => 'Image',
        'product_storebalance' => 'StoreBalance',
        'counterparty' => 'Counterparty',
        'counterparty_legal' => 'Counterparty',
        'counterparty_metadata' => 'CounterpartyMetadata',
        'counterparty_metadata_expanded' => 'CounterpartyMetadata',
        'counterparty_metadata_minimum' => 'CounterpartyMetadata',
        'document_metadata' => 'DocumentMetadata',
        'document_metadata_expanded' => 'DocumentMetadata',
        'product_metadata' => 'Metadata',
        'product_metadata_expanded' => 'Metadata',
        'counterparty_account' => 'Account',
        'counterparty_contactperson' => 'ContactPerson',
        'counterparty_note' => 'Note',
        'event_note' => 'EventNote',
        'counterparty_file' => 'File',
        'currency' => 'Currency',
        'employee' => 'Employee',
        'employee_security' => 'EmployeeSecurity',
        'employee_role' => 'EmployeeRole',
        'group' => 'Group',
        'entity_with_extra_field' => 'Group',
        'country' => 'Country',
        'region' => 'Region',
        'tax_rate' => 'TaxRate',
        'product_folder' => 'ProductFolder',
        'processing_stage' => 'ProcessingStage',
        'processing_process' => 'ProcessingProcess',
        'processing_plan' => 'ProcessingPlan',
        'processing_plan_expanded' => 'ProcessingPlan',
        'service' => 'Service',
        'uom' => 'Uom',
        'price_type' => 'PriceType',
        'sale_platform' => 'SalePlatform',
        'store' => 'Store',
        'retail_store' => 'RetailStore',
        'cashier' => 'Cashier',
        'bundle' => 'Bundle',
        'bundle_component' => 'BundleComponent',
        'variant' => 'Variant',
        'webhook' => 'Webhook',
        'webhookstock' => 'WebhookStock',
        'thing' => 'Thing',
        'internal_order' => 'InternalOrder',
        'processing_order' => 'ProcessingOrder',
        'processing' => 'Processing',
        'customer_order' => 'CustomerOrder',
        'demand' => 'Demand',
        'move' => 'Move',
        'retail_demand' => 'RetailDemand',
        'enter' => 'Enter',
        'sales_return' => 'SalesReturn',
        'loss' => 'Loss',
        'purchase_order' => 'PurchaseOrder',
        'purchase_return' => 'PurchaseReturn',
        'prepayment_return' => 'PrepaymentReturn',
        'emission_order' => 'EmissionOrder',
        'variantcharacteristic' => 'VariantCharacteristic',
        'contract' => 'Contract',
        'task' => 'Task',
        'sales_channel' => 'SalesChannel',
        'project' => 'Project',
        'consignment' => 'Consignment',
        'expense_item' => 'ExpenseItem',
        'cash_in' => 'CashIn',
        'retail_drawer_cash_in' => 'RetailDrawerCashIn',
        'retail_drawer_cash_out' => 'RetailDrawerCashOut',
        'cash_in_operation' => 'CashInOperation',
        'cash_out' => 'CashOut',
        'cash_out_operation' => 'CashOutOperation',
        'payment_in' => 'PaymentIn',
        'payment_out' => 'PaymentOut',
        'facture_in' => 'FactureIn',
        'facture_out' => 'FactureOut',
        'company_settings' => 'CompanySettings',
        'company_settings_metadata' => 'CompanySettingsMetadata',
        'user_settings' => 'UserSettings',
        'subscription' => 'Subscription',
        'assortment_settings' => 'AssortmentSettings',
        'assortment' => 'Assortment',
        'discount' => 'Discount',
        'accumulation_discount' => 'AccumulationDiscount',
        'personal_discount' => 'PersonalDiscount',
        'special_price_discount' => 'SpecialPriceDiscount',
        'bonus_program' => 'BonusProgram',
        'bonus_transaction' => 'BonusTransaction',
        'custom_entity' => 'CustomEntity',
        'custom_entity_element' => 'CustomEntityElement',
        'commission_report_in' => 'CommissionReportIn',
        'processing_plan_folder' => 'ProcessingPlanFolder',
        'organization' => 'Organization',
        'organization_account' => 'Account',
    ];

    /**
     * Поля, которые игнорируются при сравнении.
     * Все перечисленные в спецификации помечены readOnly: true (генерируются сервером).
     * При roundtrip они могут отсутствовать в выводе SDK (генератор часто не сериализует readOnly)
     * или отличаться (например, updated/created при следующем запросе).
     */
    private const IGNORED_FIELDS = [
        'updated',
        'created',
        'accountId',
        'pathName',
        'effectiveVat',
        'effectiveVatEnabled',
        'variantsCount',
        'tobacco',
        'salesAmount',
        'bonusPoints',
        'extra_field'
    ];

    /**
     * Тест на конситентность существующих fixture-файлов и маппинга FIXTURE_MODEL_MAP .
     */
    public function testMappingAndFixtureConsistency(): void
    {
        $fixturesPath = $this->getFixturesPath();
        $this->assertDirectoryExists($fixturesPath);

        $files = glob($fixturesPath . DIRECTORY_SEPARATOR . '*.json') ?: [];
        foreach ($files as $file) {
            $base = basename($file, '.json');
            $this->assertArrayHasKey(
                $base,
                self::FIXTURE_MODEL_MAP,
                'Model mapping not found for ' . basename($file)
            );
        }
    }

    /**
     * @dataProvider fixtureProvider
     * 
     * Тест roundtrip сериализации: JSON -> Model -> JSON.
     * Проверяет, что данные не теряются и не искажаются при преобразованиях.
     * 
     * @param string $fixtureName Имя fixture файла (без расширения)
     * @param string $modelName Имя класса модели
     */
    public function testRoundtripSerialization(string $fixtureName, string $modelName): void
    {
        $fixtureFile = $fixtureName . '.json';
        
        // Проверяем наличие fixture файла
        $fixturePath = $this->getFixturesPath() . '/' . $fixtureFile;
        if (!file_exists($fixturePath)) {
            $this->markTestSkipped("Fixture file not found: {$fixtureFile}");
        }

        // Загружаем эталонный JSON
        $originalJson = $this->loadFixture($fixtureFile);
        
        // Получаем полное имя класса модели
        $modelClass = $this->getModelClass($modelName);
        
        // Проверяем существование класса модели
        if (!class_exists($modelClass)) {
            $this->markTestSkipped("Model class not found: {$modelClass}. SDK may not be generated.");
        }

        // Десериализуем JSON в объект модели
        $model = $this->deserialize($originalJson, $modelClass);
        $this->assertNotNull($model, "Failed to deserialize {$fixtureFile} to {$modelName}");

        // Сериализуем модель обратно в массив
        $serializedJson = $this->serialize($model);
        $this->assertIsArray($serializedJson, "Serialization should return array");

        // Нормализуем оба массива для сравнения
        $normalizedOriginal = $this->normalizeForComparison($originalJson);
        $normalizedSerialized = $this->normalizeForComparison($serializedJson);

        // Сравниваем с учётом эквивалентности null/отсутствия ключа и чисел 1.0/1
        $this->assertNormalizedEquals(
            $normalizedOriginal,
            $normalizedSerialized,
            "Roundtrip serialization failed for {$fixtureName}: data mismatch after deserialize->serialize"
        );
    }

    /**
     * Провайдер данных для тестов.
     * Возвращает пары [fixture_name, model_name] для каждого fixture файла.
     * 
     * @return array<string, array{string, string}>
     */
    public static function fixtureProvider(): array
    {
        $testCases = [];
        
        foreach (self::FIXTURE_MODEL_MAP as $fixtureName => $modelName) {
            $testCases[$fixtureName] = [$fixtureName, $modelName];
        }
        
        return $testCases;
    }

    /**
     * Возвращает полное имя класса модели SDK.
     * 
     * @param string $shortName Короткое имя модели (например, 'Product')
     * @return string Полное имя класса
     */
    private function getModelClass(string $shortName): string
    {
        // Namespace зависит от настроек openapi-generator
        return "OpenAPI\\Client\\Model\\{$shortName}";
    }

    /**
     * Десериализует JSON массив в объект модели.
     * Использует ObjectSerializer из сгенерированного SDK.
     * 
     * @param array<string, mixed> $data JSON данные
     * @param string $class Полное имя класса модели
     * @return object|null Объект модели или null при ошибке
     */
    private function deserialize(array $data, string $class): ?object
    {
        $serializerClass = "OpenAPI\\Client\\ObjectSerializer";
        
        if (class_exists($serializerClass) && method_exists($serializerClass, 'deserialize')) {
            try {
                // ObjectSerializer::deserialize ожидает JSON строку или stdClass
                $jsonString = json_encode($data);
                return $serializerClass::deserialize($jsonString, $class);
            } catch (\Throwable $e) {
                $this->fail("Deserialization failed: " . $e->getMessage());
            }
        }

        // Fallback: попытка создать объект напрямую
        try {
            return new $class($data);
        } catch (\Throwable $e) {
            return null;
        }
    }

    /**
     * Сериализует объект модели в массив.
     * 
     * @param object $model Объект модели
     * @return array<string, mixed> Сериализованные данные
     */
    private function serialize(object $model): array
    {
        $serializerClass = "OpenAPI\\Client\\ObjectSerializer";
        
        if (class_exists($serializerClass) && method_exists($serializerClass, 'sanitizeForSerialization')) {
            $sanitized = $serializerClass::sanitizeForSerialization($model);
            return json_decode(json_encode($sanitized), true) ?? [];
        }

        if ($model instanceof \JsonSerializable) {
            return $model->jsonSerialize();
        }

        // Fallback: преобразуем объект в массив через JSON
        return json_decode(json_encode($model), true) ?? [];
    }

    /**
     * Нормализует данные для сравнения:
     * - удаляет игнорируемые поля;
     * - для attributes[].value приводит скаляр и объект/массив к одному виду;
     * - числа 1.0 и 1 приводятся к одному виду;
     * - пустые объекты и объекты только с meta считаются одинаковыми (нормализуются в []).
     *
     * @param array<string, mixed> $data Исходные данные
     * @return array<string, mixed> Нормализованные данные
     */
    private function normalizeForComparison(array $data): array
    {
        foreach (self::IGNORED_FIELDS as $field) {
            unset($data[$field]);
        }

        foreach ($data as $key => $value) {
            if (is_array($value)) {
                $data[$key] = $this->normalizeForComparison($value);
            } elseif (is_float($value) && $value === (float) (int) $value) {
                $data[$key] = (int) $value;
            }
        }

        // Атрибуты: value может прийти скаляром (фикстура) или массивом/объектом (SDK) — приводим к скаляру
        if (isset($data['attributes']) && is_array($data['attributes'])) {
            foreach ($data['attributes'] as $i => $attr) {
                if (is_array($attr) && array_key_exists('value', $attr)) {
                    $data['attributes'][$i]['value'] = $this->normalizeAttributeValue($attr['value']);
                }
            }
        }

        // Пустой объект или только meta — нормализуем в [] для совпадения с выводом SDK
        if ($data === []) {
            return [];
        }

        ksort($data);
        return $data;
    }

    /**
     * Приводит value атрибута к скалярному виду (фикстура — скаляр, SDK часто — массив/объект).
     *
     * @param mixed $value
     * @return mixed
     */
    private function normalizeAttributeValue($value)
    {
        if (!is_array($value)) {
            return $value;
        }
        // Один элемент — скаляр (например обёртка)
        if (count($value) === 1) {
            $v = reset($value);
            return is_scalar($v) ? $v : $value;
        }
        // Ищем первый скаляр во вложенной структуре
        foreach ($value as $v) {
            if (is_scalar($v)) {
                return $v;
            }
        }
        return $value;
    }

    /**
     * Рекурсивное сравнение с учётом: отсутствие ключа = null, числа 1.0 и 1 равны.
     *
     * @param array<string, mixed> $expected
     * @param array<string, mixed> $actual
     */
    private function assertNormalizedEquals(array $expected, array $actual, string $message = ''): void
    {
        $this->sortKeysRecursive($expected);
        $this->sortKeysRecursive($actual);
        $diff = $this->diffNormalized($expected, $actual);
        if ($diff !== []) {
            $this->fail($message . "\n" . 'Differences: ' . json_encode($diff, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
        }
    }

    private function sortKeysRecursive(array &$arr): void
    {
        ksort($arr);
        foreach ($arr as $k => $v) {
            if (is_array($v)) {
                $this->sortKeysRecursive($arr[$k]);
            }
        }
    }

    /**
     * Сравнивает с учётом null/отсутствия и числового равенства 1.0 == 1.
     * Возвращает массив расхождений (пустой, если равны).
     *
     * @param mixed $expected
     * @param mixed $actual
     * @return array<int, array{path: string, expected: mixed, actual: mixed}>
     */
    private function diffNormalized($expected, $actual, string $path = ''): array
    {
        if (is_array($expected) && is_array($actual)) {
            $allKeys = array_unique(array_merge(array_keys($expected), array_keys($actual)));
            $diffs = [];
            foreach ($allKeys as $key) {
                $e = $expected[$key] ?? null;
                $a = $actual[$key] ?? null;
                if ($e === null && $a === null) {
                    continue;
                }
                $subPath = $path === '' ? $key : $path . '.' . $key;
                $diffs = array_merge($diffs, $this->diffNormalized($e, $a, $subPath));
            }
            return $diffs;
        }

        if (is_numeric($expected) && is_numeric($actual) && (float) $expected === (float) $actual) {
            return [];
        }
        if ($expected === $actual) {
            return [];
        }
        return [['path' => $path, 'expected' => $expected, 'actual' => $actual]];
    }
}

File: tests/java/assertions/src/test/java/com/lognex/test/golden/SerializationTest.java
package com.lognex.test.golden;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.lognex.test.BaseTestCase;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import ru.moysklad.remap_1_2.ApiClient;

import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Golden tests for SDK serialization/deserialization.
 * <p>
 * Roundtrip flow:
 * 1) load fixture JSON
 * 2) deserialize to model object
 * 3) serialize model back to JSON-like structure
 * 4) compare with normalized fixture
 */
class SerializationTest extends BaseTestCase {

    private static final ObjectMapper MAPPER = new ApiClient().getObjectMapper();

    private static final Map<String, String> FIXTURE_MODEL_MAP = new HashMap<>();

    static {
        FIXTURE_MODEL_MAP.put("product", "Product");
        FIXTURE_MODEL_MAP.put("product_full", "Product");
        FIXTURE_MODEL_MAP.put("product_image", "Image");
        FIXTURE_MODEL_MAP.put("product_storebalance", "StoreBalance");
        FIXTURE_MODEL_MAP.put("counterparty", "Counterparty");
        FIXTURE_MODEL_MAP.put("counterparty_legal", "Counterparty");
        FIXTURE_MODEL_MAP.put("counterparty_metadata", "CounterpartyMetadata");
        FIXTURE_MODEL_MAP.put("counterparty_metadata_expanded", "CounterpartyMetadata");
        FIXTURE_MODEL_MAP.put("counterparty_metadata_minimum", "CounterpartyMetadata");
        FIXTURE_MODEL_MAP.put("document_metadata", "DocumentMetadata");
        FIXTURE_MODEL_MAP.put("document_metadata_expanded", "DocumentMetadata");
        FIXTURE_MODEL_MAP.put("product_metadata", "Metadata");
        FIXTURE_MODEL_MAP.put("product_metadata_expanded", "Metadata");
        FIXTURE_MODEL_MAP.put("counterparty_account", "Account");
        FIXTURE_MODEL_MAP.put("counterparty_contactperson", "ContactPerson");
        FIXTURE_MODEL_MAP.put("counterparty_note", "Note");
        FIXTURE_MODEL_MAP.put("counterparty_file", "ModelFile");
        FIXTURE_MODEL_MAP.put("currency", "Currency");
        FIXTURE_MODEL_MAP.put("employee", "Employee");
        FIXTURE_MODEL_MAP.put("employee_security", "EmployeeSecurity");
        FIXTURE_MODEL_MAP.put("employee_role", "EmployeeRole");
        FIXTURE_MODEL_MAP.put("group", "Group");
        FIXTURE_MODEL_MAP.put("entity_with_extra_field", "Group");
        FIXTURE_MODEL_MAP.put("country", "Country");
        FIXTURE_MODEL_MAP.put("region", "Region");
        FIXTURE_MODEL_MAP.put("tax_rate", "TaxRate");
        FIXTURE_MODEL_MAP.put("product_folder", "ProductFolder");
        FIXTURE_MODEL_MAP.put("processing_plan_folder", "ProcessingPlanFolder");
        FIXTURE_MODEL_MAP.put("service", "Service");
        FIXTURE_MODEL_MAP.put("uom", "Uom");
        FIXTURE_MODEL_MAP.put("price_type", "PriceType");
        FIXTURE_MODEL_MAP.put("sale_platform", "SalePlatform");
        FIXTURE_MODEL_MAP.put("store", "Store");
        FIXTURE_MODEL_MAP.put("retail_store", "RetailStore");
        FIXTURE_MODEL_MAP.put("cashier", "Cashier");
        FIXTURE_MODEL_MAP.put("bundle", "Bundle");
        FIXTURE_MODEL_MAP.put("bundle_component", "BundleComponent");
        FIXTURE_MODEL_MAP.put("variant", "Variant");
        FIXTURE_MODEL_MAP.put("webhook", "Webhook");
        FIXTURE_MODEL_MAP.put("webhookstock", "WebhookStock");
        FIXTURE_MODEL_MAP.put("thing", "Thing");
        FIXTURE_MODEL_MAP.put("internal_order", "InternalOrder");
        FIXTURE_MODEL_MAP.put("customer_order", "CustomerOrder");
        FIXTURE_MODEL_MAP.put("demand", "Demand");
        FIXTURE_MODEL_MAP.put("purchase_order", "PurchaseOrder");
        FIXTURE_MODEL_MAP.put("emission_order", "EmissionOrder");
        FIXTURE_MODEL_MAP.put("event_note", "EventNote");
        FIXTURE_MODEL_MAP.put("variantcharacteristic", "VariantCharacteristic");
        FIXTURE_MODEL_MAP.put("contract", "Contract");
        FIXTURE_MODEL_MAP.put("task", "Task");
        FIXTURE_MODEL_MAP.put("sales_channel", "SalesChannel");
        FIXTURE_MODEL_MAP.put("project", "Project");
        FIXTURE_MODEL_MAP.put("consignment", "Consignment");
        FIXTURE_MODEL_MAP.put("expense_item", "ExpenseItem");
        FIXTURE_MODEL_MAP.put("cash_in", "CashIn");
        FIXTURE_MODEL_MAP.put("retail_drawer_cash_in", "RetailDrawerCashIn");
        FIXTURE_MODEL_MAP.put("cash_in_operation", "CashInOperation");
        FIXTURE_MODEL_MAP.put("cash_out", "CashOut");
        FIXTURE_MODEL_MAP.put("cash_out_operation", "CashOutOperation");
        FIXTURE_MODEL_MAP.put("facture_in", "FactureIn");
        FIXTURE_MODEL_MAP.put("facture_out", "FactureOut");
        FIXTURE_MODEL_MAP.put("company_settings", "CompanySettings");
        FIXTURE_MODEL_MAP.put("company_settings_metadata", "CompanySettingsMetadata");
        FIXTURE_MODEL_MAP.put("user_settings", "UserSettings");
        FIXTURE_MODEL_MAP.put("subscription", "Subscription");
        FIXTURE_MODEL_MAP.put("assortment_settings", "AssortmentSettings");
        FIXTURE_MODEL_MAP.put("assortment", "Assortment");
        FIXTURE_MODEL_MAP.put("discount", "Discount");
        FIXTURE_MODEL_MAP.put("accumulation_discount", "AccumulationDiscount");
        FIXTURE_MODEL_MAP.put("personal_discount", "PersonalDiscount");
        FIXTURE_MODEL_MAP.put("special_price_discount", "SpecialPriceDiscount");
        FIXTURE_MODEL_MAP.put("bonus_program", "BonusProgram");
        FIXTURE_MODEL_MAP.put("bonus_transaction", "BonusTransaction");
        FIXTURE_MODEL_MAP.put("custom_entity", "CustomEntity");
        FIXTURE_MODEL_MAP.put("custom_entity_element", "CustomEntityElement");
        FIXTURE_MODEL_MAP.put("commission_report_in", "CommissionReportIn");
    }

    private static final Set<String> IGNORED_FIELDS = new HashSet<>();

    static {
        IGNORED_FIELDS.add("updated");
        IGNORED_FIELDS.add("created");
        IGNORED_FIELDS.add("accountId");
        IGNORED_FIELDS.add("pathName");
        IGNORED_FIELDS.add("effectiveVat");
        IGNORED_FIELDS.add("effectiveVatEnabled");
        IGNORED_FIELDS.add("variantsCount");
        IGNORED_FIELDS.add("tobacco");
        IGNORED_FIELDS.add("salesAmount");
        IGNORED_FIELDS.add("bonusPoints");
        IGNORED_FIELDS.add("extra_field");//check objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    /**
     * Тест на конситентность существующий fixture-файлов и маппинга FIXTURE_MODEL_MAP.
     */
    @Test
    void testMappingAndFixtureConsistency() throws URISyntaxException {
        final File[] files = getFixturesPath().toFile().listFiles((dir, name) -> name.endsWith(".json"));
        for (File file : files) {
            assertTrue(FIXTURE_MODEL_MAP.containsKey(file.getName().replace(".json", "")),
                    "Model mapping not found for " + file.getName());
        }
    }

    /**
     * Тест roundtrip сериализации: JSON -> Model -> JSON.
     * Проверяет, что данные не теряются и не искажаются при преобразованиях.
     *
     * @param fixtureName Имя fixture файла (без расширения)
     * @param modelName   Имя класса модели
     */
    @ParameterizedTest(name = "{0}")
    @MethodSource("fixtureProvider")
    void testRoundtripSerialization(String fixtureName, String modelName) throws URISyntaxException {
        String fixtureFile = fixtureName + ".json";
        Path fixturePath = getFixturesPath().resolve(fixtureFile);
        Assertions.assertTrue(Files.exists(fixturePath), "Fixture file not found: " + fixtureFile);

        Map<String, Object> originalJson = loadFixture(fixtureFile);
        String modelClass = getModelClass(modelName);

        Class<?> clazz = loadClass(modelClass);
        assertNotNull(clazz, "Model class not found: " + modelClass + ". SDK may not be generated.");

        Object model = deserialize(originalJson, clazz);
        Assertions.assertNotNull(model, "Failed to deserialize " + fixtureFile + " to " + modelName);

        Map<String, Object> serializedJson = serialize(model);

        Map<String, Object> normalizedOriginal = normalizeForComparison(originalJson);
        Map<String, Object> normalizedSerialized = normalizeForComparison(serializedJson);

        assertNormalizedEquals(
                normalizedOriginal,
                normalizedSerialized,
                "Roundtrip serialization failed for " + fixtureName + ": data mismatch after deserialize->serialize"
        );
    }

    private static Stream<org.junit.jupiter.params.provider.Arguments> fixtureProvider() {
        return FIXTURE_MODEL_MAP.entrySet().stream()
                .map(e -> org.junit.jupiter.params.provider.Arguments.of(e.getKey(), e.getValue()));
    }

    private String getModelClass(String shortName) {
        return "ru.moysklad.remap_1_2.model." + shortName;
    }

    private Class<?> loadClass(String className) {
        try {
            return Class.forName(className);
        } catch (ClassNotFoundException e) {
            return null;
        }
    }

    private Object deserialize(Map<String, Object> data, Class<?> clazz) {
        try {
            return MAPPER.convertValue(data, clazz);
        } catch (Exception primary) {
            try {
                Constructor<?> constructor = clazz.getDeclaredConstructor(Map.class);
                constructor.setAccessible(true);
                return constructor.newInstance(data);
            } catch (Exception ignored) {
                return null;
            }
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> serialize(Object model) {
        try {
            Method m = model.getClass().getMethod("toMap");
            Object out = m.invoke(model);
            if (out instanceof Map) {
                return (Map<String, Object>) out;
            }
        } catch (Exception ignored) {
            // Fall back to Jackson conversion if SDK-specific API is absent.
        }
        return (Map<String, Object>) MAPPER.convertValue(model, new TypeReference<Object>() {});
    }

    private Map<String, Object> normalizeForComparison(Map<String, Object> data) {
        Map<String, Object> normalized = new LinkedHashMap<>();

        for (Map.Entry<String, Object> entry : data.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();

            if (IGNORED_FIELDS.contains(key) /*|| value == null*/) {
                continue;
            }
            normalized.put(key, normalizeValue(value));
        }

        Object attributes = normalized.get("attributes");
        if (attributes instanceof List) {
            List<Object> fixed = new ArrayList<>();
            List<?> l = (List<?>) attributes;
            for (Object item : l) {
                if (item instanceof Map) {
                    Map<String, Object> attr = castMap((Map<?, ?>) item);
                    if (attr.containsKey("value")) {
                        attr.put("value", normalizeAttributeValue(attr.get("value")));
                    }
                    fixed.add(attr);
                } else {
                    fixed.add(item);
                }
            }
            normalized.put("attributes", fixed);
        }

        if (normalized.isEmpty()) {
            return new LinkedHashMap<>();
        }

        return new TreeMap<>(normalized);
    }

    private Object normalizeValue(Object value) {
        if (value instanceof Map) {
            return normalizeForComparison(castMap((Map<?, ?>) value));
        }
        if (value instanceof List) {
            List<Object> out = new ArrayList<>();
            for (Object item : ((List) value)) {
                out.add(normalizeValue(item));
            }
            return out;
        }
        if (value instanceof Double && (Double) value == Math.rint((Double) value)) {
            return (int) ((Double) value).doubleValue();
        }
        if (value instanceof Float && ((Float) value) == Math.rint((Float) value)) {
            return (int) ((Float) value).floatValue();
        }
        return value;
    }

    private Object normalizeAttributeValue(Object value) {
        if (!(value instanceof List)) {
            return value;
        }
        List<?> list = (List<?>) value;
        if (list.size() == 1) {
            Object single = list.get(0);
            return isScalar(single) ? single : value;
        }
        for (Object v : list) {
            if (isScalar(v)) {
                return v;
            }
        }
        return value;
    }

    private static boolean isScalar(Object value) {
        return value instanceof String
               || value instanceof Number
               || value instanceof Boolean
               || value instanceof Character;
    }

    private void assertNormalizedEquals(Map<String, Object> expected, Map<String, Object> actual, String message) {
        Map<String, Object> expectedSorted = sortKeysRecursive(expected);
        Map<String, Object> actualSorted = sortKeysRecursive(actual);
        List<Map<String, Object>> diff = diffNormalized(expectedSorted, actualSorted, "");
        if (!diff.isEmpty()) {
            Assertions.fail(message + "\nDifferences: " + toPrettyJson(diff));
        }
    }

    private Map<String, Object> sortKeysRecursive(Map<String, Object> input) {
        Map<String, Object> sorted = new TreeMap<>();
        for (Map.Entry<String, Object> e : input.entrySet()) {
            Object value = e.getValue();
            if (value instanceof Map) {
                sorted.put(e.getKey(), sortKeysRecursive(castMap((Map<?, ?>) value)));
            } else if (value instanceof List) {
                List<Object> mapped = new ArrayList<>();
                for (Object item : ((List) value)) {
                    if (item instanceof Map) {
                        mapped.add(sortKeysRecursive(castMap((Map<?, ?>) item)));
                    } else {
                        mapped.add(item);
                    }
                }
                sorted.put(e.getKey(), mapped);
            } else {
                sorted.put(e.getKey(), value);
            }
        }
        return sorted;
    }

    private List<Map<String, Object>> diffNormalized(Object expected, Object actual, String path) {
        if (expected instanceof Map && actual instanceof Map) {
            Set<String> allKeys = new LinkedHashSet<>();
            allKeys.addAll(castMap((Map<?, ?>) expected).keySet());
            allKeys.addAll(castMap((Map<?, ?>) actual).keySet());

            List<Map<String, Object>> diffs = new ArrayList<>();
            for (String key : allKeys) {
                Object e = castMap((Map<?, ?>) expected).getOrDefault(key, "заглушка для ключа, несуществующего в fixture");
                Object a = castMap((Map<?, ?>) actual).getOrDefault(key, null);
                if (e == null && a == null) {
                    continue;
                }
                String subPath = path.isEmpty() ? key : path + "." + key;
                diffs.addAll(diffNormalized(e, a, subPath));
            }
            return diffs;
        }

        if (expected instanceof List && actual instanceof List) {
            List<?> expectedList = (List<?>) expected;
            List<?> actualList = (List<?>) actual;
            int max = Math.max(expectedList.size(), actualList.size());
            List<Map<String, Object>> diffs = new ArrayList<>();
            for (int i = 0; i < max; i++) {
                Object e = i < expectedList.size() ? expectedList.get(i) : null;
                Object a = i < actualList.size() ? actualList.get(i) : null;
                if (e == null && a == null) {
                    continue;
                }
                String subPath = path + "[" + i + "]";
                diffs.addAll(diffNormalized(e, a, subPath));
            }
            return diffs;
        }

        if (bothNumericAndEqual(expected, actual)) {
            return Collections.emptyList();
        }
        if (Objects.equals(expected, actual)) {
            return Collections.emptyList();
        }

        Map<String, Object> mismatch = new LinkedHashMap<>();
        mismatch.put("path", path);
        mismatch.put("expected", expected);
        mismatch.put("actual", actual);
        return Collections.singletonList(mismatch);
    }

    private static boolean bothNumericAndEqual(Object expected, Object actual) {
        if (!(expected instanceof Number) || !(actual instanceof Number)) {
            return false;
        }
        double e = ((Number) expected).doubleValue();
        double a = ((Number) actual).doubleValue();
        return Double.compare(e, a) == 0;
    }

    private static Map<String, Object> castMap(Map<?, ?> map) {
        Map<String, Object> out = new LinkedHashMap<>();
        for (Map.Entry<?, ?> e : map.entrySet()) {
            out.put(String.valueOf(e.getKey()), e.getValue());
        }
        return out;
    }

    private String toPrettyJson(Object value) {
        try {
            return MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(value);
        } catch (Exception e) {
            return String.valueOf(value);
        }
    }
}
